cg/meta/workflow/mutant/quality_controller/models.py

```diff
@@ -0,0 +1,89 @@
+from typing import Annotated, Any
+from pydantic import BaseModel, BeforeValidator, Field, ValidationError, ConfigDict
+from cg.store.models import Sample
+
+
+# Validators
+def str_to_bool(value: str) -> bool:
+    if value == "TRUE":
+        return True
+    elif value == "FALSE":
+        return False
+    raise ValidationError(f"String {value} cannot be turned to bool.")
+
+
+# Models
+class SampleResults(BaseModel):
+    sample_name: str = Field(alias="Sample")
+    selection: str = Field(alias="Selection")
+    region_code: str = Field(alias="Region Code")
+    ticket: int = Field(alias="Ticket")
+    pct_n_bases: float = Field(alias="%N_bases")
+    pct_10x_coverage: float = Field(alias="%10X_coverage")
+    qc_pass: Annotated[bool, BeforeValidator(str_to_bool)] = Field(alias="QC_pass")
+    lineage: str = Field(alias="Lineage")
+    pangolin_data_version: str = Field(alias="Pangolin_data_version")
+    voc: str = Field(alias="VOC")
+    mutations: str = Field(alias="Mutations")
+
+
+class MutantPoolSamples(BaseModel):
+    samples: list[Sample]
+    external_negative_control: Sample
+    internal_negative_control: Sample
+
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+
+class QualityMetrics(BaseModel):
+    results: dict[str, SampleResults]
+    pool: MutantPoolSamples
+
+
+class SampleQualityResults(BaseModel):
+    sample_id: str
+    passes_qc: bool
+    passes_reads_threshold: bool
+    passes_mutant_qc: bool | None = None
+
+
+class SamplesQualityResults(BaseModel):
+    internal_negative_control: SampleQualityResults
+    external_negative_control: SampleQualityResults
+    samples: list[SampleQualityResults]
+
+    @property
+    def total_samples_count(self) -> int:
+        return len(self.samples)
+
+    @property
+    def passed_samples_count(self) -> int:
+        samples_pass_qc: list[bool] = [sample_result.passes_qc for sample_result in self.samples]
+        return sum(samples_pass_qc)
+
+    @property
+    def failed_samples_count(self) -> int:
+        return self.total_samples_count - self.passed_samples_count
+
+
+class CaseQualityResult(BaseModel):
+    passes_qc: bool
+    internal_negative_control_passes_qc: bool
+    external_negative_control_passes_qc: bool
+    samples_pass_qc: bool
+
+
+class QualityResult(BaseModel):
+    case_quality_result: CaseQualityResult
+    samples_quality_results: SamplesQualityResults
+    summary: str
+
+    @property
+    def passes_qc(self) -> bool:
+        return self.case_quality_result.passes_qc
+
+
+class Report(BaseModel):
```

Comment:
As Isak suggested, some class names might be too broad and could lead to confusion when adding new QC controllers.

For example, `QualityResult` implies a QC analysis result, but you'd need to check the file or import to understand its exact use. It's an extreme example 😆, but a non-descriptive class name can be frustrating when you're trying to find the right one

Reply:
I definitely understand what you mean, but I am unsure of where the line should be: So should all of these models be called `MutantModelName` (i.e. `MutantReport`, `MutantQualityResult`, etc.) to indicate that they refer to mutant qc? Wouldn't it be decriptive enough that one imports these models from `mutant.quality_controller.models`?
If the answer is no to both questions, which models should have a `Mutant` prefix and which are ok having general names?

Reply:
Could be difficult to trace the line but I think we just have to try be as descriptive as possible with our model names to not incur in something like this, for example:

```
from delivery_report import Report as DeliveryReport
from mutant import Report as MutantReport
```

It's OK to use namespaces but these reports have such a different purpose that's better to avoid the ambiguity imo.

Since we don't have anything written in stone here, I would at least rename the Report model. Could also be a good idea to have another look at the other names, but I'll leave it up to you.

Reply:
Upon discussion with @diitaz93, let's add Mutant to the classes that are used outside of quality_controller. So MutantReport and MutantQualityResult.